File: pkg/codegen/testing/test/testdata/components-pp/nodejs/exampleComponent.ts
import * as pulumi from "@pulumi/pulumi";
import * as random from "@pulumi/random";
import { SimpleComponent } from "./simpleComponent";

interface ExampleComponentArgs {
    /**
     * A simple input
     */
    input: pulumi.Input<string>,
    /**
     * The main CIDR blocks for the VPC
     * It is a map of strings
     */
    cidrBlocks: pulumi.Input<Record<string, pulumi.Input<string>>>,
    /**
     * GitHub app parameters, see your github app. Ensure the key is the base64-encoded `.pem` file (the output of `base64 app.private-key.pem`, not the content of `private-key.pem`).
     */
    githubApp: {
        id?: pulumi.Input<string>,
        keyBase64?: pulumi.Input<string>,
        webhookSecret?: pulumi.Input<string>,
    },
    /**
     * A list of servers
     */
    servers: {
        name?: pulumi.Input<string>,
    }[],
    /**
     * A map between for zones
     */
    deploymentZones: Record<string, {
        zone?: pulumi.Input<string>,
    }>,
    ipAddress: pulumi.Input<number[]>,
}

export class ExampleComponent extends pulumi.ComponentResource {
    public result: pulumi.Output<string>;
    constructor(name: string, args: ExampleComponentArgs, opts?: pulumi.ComponentResourceOptions) {
        super("components:index:ExampleComponent", name, args, opts);
        const password = new random.RandomPassword(`${name}-password`, {
            length: 16,
            special: true,
            overrideSpecial: args.input,
        }, {
            parent: this,
        });

        const githubPassword = new random.RandomPassword(`${name}-githubPassword`, {
            length: 16,
            special: true,
            overrideSpecial: args.githubApp.webhookSecret,
        }, {
            parent: this,
        });

        // Example of iterating a list of objects
        const serverPasswords: random.RandomPassword[] = [];
        for (const range = {value: 0}; range.value < args.servers.length; range.value++) {
            serverPasswords.push(new random.RandomPassword(`${name}-serverPasswords-${range.value}`, {
                length: 16,
                special: true,
                overrideSpecial: args.servers[range.value].name,
            }, {
            parent: this,
        }));
        }

        // Example of iterating a map of objects
        const zonePasswords: random.RandomPassword[] = [];
        for (const range of Object.entries(args.deploymentZones).map(([k, v]) => ({key: k, value: v}))) {
            zonePasswords.push(new random.RandomPassword(`${name}-zonePasswords-${range.key}`, {
                length: 16,
                special: true,
                overrideSpecial: range.value.zone,
            }, {
            parent: this,
        }));
        }

        const simpleComponent = new SimpleComponent(`${name}-simpleComponent`, {
            parent: this,
        });

        this.result = password.result;
        this.registerOutputs({
            result: password.result,
        });
    }
}

File: pkg/codegen/testing/test/testdata/components-pp/dotnet/ExampleComponent.cs
using System.Collections.Generic;
using System.Linq;
using Pulumi;
using Random = Pulumi.Random;

namespace Components
{
    public class ExampleComponentArgs : global::Pulumi.ResourceArgs
    {
        public class DeploymentZonesArgs : global::Pulumi.ResourceArgs
        {
            [Input("zone")]
            public Input<string>? Zone { get; set; }
        }

        public class GithubAppArgs : global::Pulumi.ResourceArgs
        {
            [Input("id")]
            public Input<string>? Id { get; set; }
            [Input("keyBase64")]
            public Input<string>? KeyBase64 { get; set; }
            [Input("webhookSecret")]
            public Input<string>? WebhookSecret { get; set; }
        }

        public class ServersArgs : global::Pulumi.ResourceArgs
        {
            [Input("name")]
            public Input<string>? Name { get; set; }
        }

        /// <summary>
        /// A simple input
        /// </summary>
        [Input("input")]
        public Input<string> Input { get; set; } = null!;
        /// <summary>
        /// The main CIDR blocks for the VPC
        /// It is a map of strings
        /// </summary>
        [Input("cidrBlocks")]
        public InputMap<string> CidrBlocks { get; set; } = null!;
        /// <summary>
        /// GitHub app parameters, see your github app. Ensure the key is the base64-encoded `.pem` file (the output of `base64 app.private-key.pem`, not the content of `private-key.pem`).
        /// </summary>
        [Input("githubApp")]
        public GithubAppArgs GithubApp { get; set; } = null!;
        /// <summary>
        /// A list of servers
        /// </summary>
        [Input("servers")]
        public ServersArgs[] Servers { get; set; } = null!;
        /// <summary>
        /// A map between for zones
        /// </summary>
        [Input("deploymentZones")]
        public Dictionary<string, DeploymentZonesArgs> DeploymentZones { get; set; } = null!;
        [Input("ipAddress")]
        public InputList<int> IpAddress { get; set; } = null!;
    }

    public class ExampleComponent : global::Pulumi.ComponentResource
    {
        [Output("result")]
        public Output<string> Result { get; private set; }
        public ExampleComponent(string name, ExampleComponentArgs args, ComponentResourceOptions? opts = null)
            : base("components:index:ExampleComponent", name, args, opts)
        {
            var password = new Random.RandomPassword($"{name}-password", new()
            {
                Length = 16,
                Special = true,
                OverrideSpecial = args.Input,
            }, new CustomResourceOptions
            {
                Parent = this,
            });

            var githubPassword = new Random.RandomPassword($"{name}-githubPassword", new()
            {
                Length = 16,
                Special = true,
                OverrideSpecial = args.GithubApp.WebhookSecret,
            }, new CustomResourceOptions
            {
                Parent = this,
            });

            // Example of iterating a list of objects
            var serverPasswords = new List<Random.RandomPassword>();
            for (var rangeIndex = 0; rangeIndex < args.Servers.Length; rangeIndex++)
            {
                var range = new { Value = rangeIndex };
                serverPasswords.Add(new Random.RandomPassword($"{name}-serverPasswords-{range.Value}", new()
                {
                    Length = 16,
                    Special = true,
                    OverrideSpecial = args.Servers[range.Value].Name,
                }, new CustomResourceOptions
                {
                    Parent = this,
                }));
            }
            // Example of iterating a map of objects
            var zonePasswords = new List<Random.RandomPassword>();
            foreach (var range in args.DeploymentZones.Select(pair => new { pair.Key, pair.Value }))
            {
                zonePasswords.Add(new Random.RandomPassword($"{name}-zonePasswords-{range.Key}", new()
                {
                    Length = 16,
                    Special = true,
                    OverrideSpecial = range.Value.Zone,
                }, new CustomResourceOptions
                {
                    Parent = this,
                }));
            }
            var simpleComponent = new Components.SimpleComponent($"{name}-simpleComponent", new ComponentResourceOptions
            {
                Parent = this,
            });

            this.Result = password.Result;

            this.RegisterOutputs(new Dictionary<string, object?>
            {
                ["result"] = password.Result,
            });
        }
    }
}
